typeclass-experiments/src/Common/Types.hs

```diff
@@ -0,0 +1,175 @@
+                rest is either Nil or a tyList of Constrs -}
+  | VarP Text {- This isn't a type variable. Although it is used to represent them in certain contexts,
+                 it is also used more generally to refer to any "hole" in a pattern to which another pattern
+                 may be substituted. We could have separate constr for type variables but it doesn't appear to be
+                 necessary at this time. -}
+  | RefP Pat {- still unclear on how to handle these here -}
+  | AppP Pat Pat {- Pattern for Type applications -}
+  | {- This last one is a bit special. This represents a complete type declaration.
+
+       The first Pat should be instantiated to `Name l` where l is a concrete name.
+
+       The second Pat should be instantiated to a Pat-List (using :*/Nil) which only contains Names.
+
+       The final Pat should be instantiated to a Pat body.
+
+       In some languages, parts of this may be ignored. E.g. in Rust the type name doesn't matter (we use the constr name of the
+       outermost inner sum for constructing types). -}
+    DecP Pat Pat Pat
+  deriving (Show, Eq, Ord)
+
+infixr 5 :*
+
+-- Pattern synonyms. These VASTLY improve readability
+
+pattern List :: Pat -> Pat
+pattern List t = AppP ListP t
+
+pattern Maybe :: Pat -> Pat
+pattern Maybe t = AppP MaybeP t
+
+pattern Map :: Pat -> Pat -> Pat
+pattern Map k v = AppP (AppP MapP k) v
+
+pattern Either :: Pat -> Pat -> Pat
+pattern Either l r = AppP (AppP EitherP l) r
+
+pattern RecFields :: Pat -> Pat -> Pat -> Pat
+pattern RecFields l x xs = RecP ((l := x) :* xs)
+
+pattern ProdArgs :: Pat -> Pat -> Pat
+pattern ProdArgs x xs = ProdP (x :* xs)
+
+pattern RecBody :: Pat -> Pat -> Pat
+pattern RecBody constr fields = SumP (constr := RecP fields :* Nil)
+
+pattern ProdBody :: Pat -> Pat -> Pat
+pattern ProdBody constr fields = SumP ((constr := ProdP fields) :* Nil)
+
+pattern Rec :: Pat -> Pat -> Pat -> Pat -> Pat
+pattern Rec tyName tyVars constr fields =
+  DecP tyName tyVars (SumP (constr := RecP fields :* Nil))
+
+pattern Product :: Pat -> Pat -> Pat -> Pat -> Pat
+pattern Product tyName tyVars constr args =
+  DecP tyName tyVars (SumP (constr := ProdP args :* Nil))
+
+pattern SumBody :: Pat -> Pat -> Pat -> Pat
+pattern SumBody l x xs = SumP ((l := x) :* xs)
+
+pattern Sum :: Pat -> Pat -> Pat -> Pat
+pattern Sum tyName tyVars pat =
+  DecP tyName tyVars pat
+
+pattern Void :: Pat
+pattern Void = SumP Nil
+
+{- Utility functions. Turn a list of types into a product/record/sum type.
+-}
+toProd :: [Pat] -> Pat
+toProd = ProdP . foldr (:*) Nil
+
+toRec :: [Pat] -> Pat
+toRec = RecP . foldr (:*) Nil
+
+toSum :: [Pat] -> Pat
+toSum = SumP . foldr (:*) Nil
+
+{- Utility functions. Turn a list of types into a product/record/sum type.
+-}
+defToPat :: Ty.TyDef -> Pat
+defToPat Ty.TyDef {..} = DecP (Name tyDefName) vars $ case tyDefBody of
+  Ty.Sum constrs -> toSum . NE.toList . fmap goConstr $ constrs
```

Comment:
This just maps TyDef to Pat? How are they different?

Reply:
It has more holes and it's flat. E.g. You can put a `VarP "x"` _anywhere_ in a pattern (more or less), which gives us a first class representation of things like "This is a list of one element" (which is relevant when generating rust types b/c turning `data F = F Int` into `enum F { F(Int) }` is super unidiomatic) 

I mean generally it's the difference between a pattern and an expression? I don't think I can do justice to the whole point of patterns in a reply here. If Haskell were more advanced we could just use Haskell patterns, but they aren't first class so we can't pass them around, which we need to do for constraint solving at least.

Reply:
If you really want to use that Control.Unify library I can give it a go but I think it would just make a lot of our code more confusing than it needs to be. 

`Pat` is NOT isomorphic to `Ty`. You can have something like `DecP (Name "Name") Nil (VarP "body")` and you _can't_ turn that into a `Ty` without instantiating the `VarP "body"` to some concrete TyBody. Maybe that helps explain why there are "more holes" and why that's useful.